### DISCUSSION TITLE
React component being rendered multiple times

### DISCUSSION THREAD

**Discussion by AshishSolanki4172** [community/community#123618] [Programming Help]
## Topic Area
React, Search Functionality, Delete Functionality, Popup

## Feature Area
- Building a search functionality with some fields
- Displaying search results in a table
- Implementing a delete functionality for each row item in the table
- Displaying a popup on click of the delete button to confirm deletion

## Issues
- The popup component is being called twice on click of the delete button
- The issue occurs based on the number of results currently visible in the table
- The popup log doesn't appear again on clicking the delete button for the second time

## Body
I just started with React and was trying to build some functionalities to understand React better. Here, I am building a search functionality with some fields. On searching, I get a table with the results. In the table, I have a delete field functionality for each row item. On clicking the delete button, a popup should open asking if the user wants to delete it or not, and store the response for further actions.

Here, on click of the delete button, it looks like the popup component is being called twice (maybe based on the number of results currently visible based on the number of log statements printed).

My table component code:

```jsx
function DataTable(props) {
    const [searchedData, setsearchedData] = useState({});
    const [openPopup, setOpenPopup] = useState(false);
    const keys = ["Customer Name", "Contact Number", "Contract Number", "Liscence Key"];

    useEffect(() => {
        *getting the data from API and storing in searchData state*
    }, [props.data]);

    function popupResponse(value) {
        setOpenPopup(false);
        if (value) {
            console.log("value is true");
        } else {
            console.log("value is false");
        }
    }

    const deleteKey = useCallback(() => {
        setOpenPopup(true);
        console.log("in delete key");
    }, [setOpenPopup]);

    return (
        <div className="data-table ">
            { Object.keys(searchedData).length > 0 ?
                <div>
                    <table className="data-table__table">
                        <thead>
                            <tr>
                                {keys.map((heading, key) => {
                                    return (
                                        <th key={key + heading}>{heading}</th>
                                    );
                                })}
                            </tr>
                        </thead>
                        <tbody>
                            {searchedData.map((data, key) => <RowData rowData={data} key={key} uniqueKey={key} openPopup={openPopup} deleteLiscense={deleteKey}/>)}
                        </tbody>
                    </table>

                    /** triggering the popup***/
                    {openPopup && <PopUp className="data-table__popup" open={openPopup} close={popupResponse}/>}
                </div>
                : <div className="data-table__no-results">No results found.</div>
            }
        </div>
    );
}
```

RowData component (call deleteLiscense function to set the state of popup to open):

```jsx
const RowData = ({rowData, uniqueKey, deleteLiscense}) => {
    const values = Object.values(rowData);

    return (
        <>
            <tr key={uniqueKey}>
                {values.map((value) => <td className="" key={uniqueKey + value}>{value}</td>)}
                <td className="" key={uniqueKey}>
                    <button key={uniqueKey} type="button" onClick={() => deleteLiscense}>Delete key</button>
                </td>
            </tr>
        </>
    );
}
```

PopUp component:

```jsx
function PopUp ({open, close}) {
    if (!open) return null;

    const response = (value) => {
        close(value)
    }

    console.log("in popup");

    return (
        <div className="data-table__popup-container">
            <div className="data-table__popup-container--heading"> Are you sure you want to delete the Liscence Key?</div>
            <div className="data-table__popup-container--button-wrapper">
                <button className="button-yes" onClick={() => response(true)}>Yes</button>
                <button className="button-no" onClick={() => response(false)}>No</button>
            </div>
        </div>
    );
}

export default PopUp;
```

I tried using the callback hook on the deleteKey functionality, thinking maybe it is getting redefined for each delete key button present in the table row, which in return getting called the same number of times as the number of row data present, triggering the popup component. But it didn't resolve the issue.

More observations: On clicking any of the delete keys for the second time, the popup log doesn't appear again.

Thanks

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by Pulkit1822** [comment]
**Resolving Repeated Rendering Issue in React Component**
=========================================================

When you encounter a situation where a popup component keeps appearing repeatedly due to the main component refreshing multiple times, it's crucial to address this issue to ensure a smooth user experience. Here's a detailed guide on how to troubleshoot and resolve this problem:

### Insights:
1. **Issue Description:** The popup component is being rendered multiple times within the DataTable component, leading to repetitive appearances.
2. **Root Cause:** The continuous refresh of the DataTable component triggers the popup to show up repeatedly, impacting the user interface.

### Suggestions for Resolution:
1. **Control Popup Rendering:** To prevent the popup from showing up multiple times, consider controlling its visibility from a broader perspective within the application.
2. **Utilize React Portals:** Explore using React Portals to render the popup component outside the main component hierarchy, ensuring it doesn't get affected by frequent refreshes.
3. **Optimize State Changes:** Review the functions responsible for state changes to avoid unnecessary refreshes that trigger the popup display.
4. **Avoid Unnecessary Refreshes:** Ensure that functions and components are optimized to minimize unnecessary re-renders and maintain a smooth user experience.

### Code Enhancement:
Here's an enhanced version of the DataTable component with optimized rendering to prevent the popup from appearing repeatedly:

```jsx
// Updated DataTable component with optimized rendering
const DataTable = (props) => {
  const [searchedData, setSearchedData] = useState([]);
  const [openPopup, setOpenPopup] = useState(false);
  const keys = ['Customer Name', 'Contact Number', 'Contract Number', 'License Key'];

  useEffect(() => {
    // Fetch data from API and store in searchedData state
    setSearchedData(props.data);
  }, [props.data]);

  const popupResponse = (value) => {
    setOpenPopup(false);
    if (value) {
      console.log('Value is true');
    } else {
      console.log('Value is false');
    }
  };

  const deleteKey = useCallback(() => {
    setOpenPopup(true);
    console.log('In delete key');
  }, []);

  return (
    <div className="data-table">
      {searchedData.length > 0 ? (
        <table className="data-table__table">
          <thead>
            <tr>
              {keys.map((heading, key) => (
                <th key={key}>{heading}</th>
              ))}
              <th></th>
            </tr>
          </thead>
          <tbody>
            {searchedData.map((data, key) => (
              <RowData
                rowData={data}
                key={key}
                uniqueKey={key}
                deleteKey={() => deleteKey(key)}
              />
            ))}
          </tbody>
        </table>
      ) : (
        <div className="data-table__no-results">No results found.</div>
      )}

      {openPopup && (
        <PopUp className="data-table__popup" open={openPopup} close={popupResponse} />
      )}
    </div>
  );
};
```

By implementing these suggestions and optimizing the component rendering, you can prevent the popup from appearing repeatedly and ensure a more efficient user interface.

**Mark as Correct Answer ✅**
-------------------------

If you found this solution helpful, please mark me as the correct answer. Feel free to reach out if you need further assistance with optimizing your React components!
